api: Allow hackers to create dev API keys with CORS

victorges opened the pull request:
<!-------------------------------------------------------------------------
 | Thanks for send a pull request! 🎉
 | First, please make sure you familiar with the contribution guidelines
 | https://github.com/livepeer/livepeer.com/blob/master/CONTRIBUTING.md
 -------------------------------------------------------------------------->

**What does this pull request do? Explain your changes. (required)**

This creates a hopefully complete solution for all the CORS problems we have been having with hackers.
During hackathons, the hackers very frequently don't have the time nor the expertise to create backends
for their applications. In those cases, they frequently encounter issues when deploying their applications
for the first time to a hosted platform, which is something they normally do in the last minute of hackathons
and then render some broken demos and frustration on hackers and our side with the CORS FAQs.

This will also reduce the complexity for getting started with Livepeer, by not requiring hackers to run some
[dummy proxies like this](https://github.com/victorges/livepeer-web-api-proxy) with in the end just expose
their API/accounts anyway...

I also aimed for an experience in the dashboard that will not only allow them to create API keys that are 
enabled for CORS usage, but also provide some helpful/educational text so they can understand what they
are doing when they enable an API key for CORS access.

Finally, to reduce the impact of hackers enabling CORS on their keys and just forgetting about it until their
app goes live, I've also limited the power that an API key has by default if it has CORS enabled (but if they
(hopefully) know what they''re doing they can also create it with full access). This will make sure that the
API key won't have powers like listing all streams/assets that exist in the account nor deleting resources
they don't already know the ID for. I tried only allowing access to creation APIs and APIs that require the
caller to know at least one resource ID, like an asset or a stream (meaning they probably got that ID from
somewhere else like a stream that they just created or saved in cookies).

Made sure that all APIs called by the Video NFT SDK are also allowed for those restricted keys.

**Specific updates (required)**

Implementing this was a little complex since it require some structural changes in our authorization logic. 
To allow changing the CORS allow list depending on the API key, we needed to parse and fetch the API key
before the CORS middleware. Also didn't want to move CORS all the way to our API handlers, so 
the first step was to:
 - Split authorization middleware in 2: an `authenticator` (global), which parses/validates the credentials, and an
 `authorizer`, which checks if those credentials give access to the currently called API.

Apart from that refactor:
 - Created param in the API to give them CORS access to specified allowed origins;
 - Restricted access of those API keys by default, only allowing access to APIs that have `allowCorsApiKey: true` in the authorization parameters
 - Created another param in the API keys to give them full access even with CORS configured
 - Updated API key creation dialog in the dashboard to set the above params in API keys

Also fixed some other CORS issues in the API:
 - Remove the static list of allowed origins, which contained all of prod, staging and development values, and receive them as a CLI arg/env instead
 - Remove CORS middlewares from VOD. They were never used anyway since afaik the outer/global middleware would already return the CORS stuff in the response (at least in the complex OPTIONS requests). Even if that's not the case, it is not necessary anymore anyway.
 - Allow any origin to access specifically the `/api/asset/upload/:url` API, since that's one that we expect frontends to call with no need for any kind of authorization.

## -

- **How did you test each of these updates (required)**
 - Created API keys with some combination of allowed origins like localhost, localhost:3000, example.com etc
 - Called different APIs with those keys, checking that the restricted/full access works and also that the CORS headers in the response come OK

Kinda still need testing if OPTIONS requests will work as I hope. My concern here is that the browser won't include
the `Authorization` header in the request and then the server doesn't know what is the API key to know what origins
it can come from.

**Does this pull request close any open issues?**
Implements #986 

**Screenshots (optional):**
<img width="474" alt="Screen Shot 2022-04-04 at 18 39 28" src="https://user-images.githubusercontent.com/1613383/161636693-69921fe3-b216-4416-80e8-1d21d751991c.png">
<img width="496" alt="Screen Shot 2022-04-04 at 18 40 12" src="https://user-images.githubusercontent.com/1613383/161636799-6196a11f-bb90-4da5-9ce7-0b31fbd5df38.png">
<img width="479" alt="Screen Shot 2022-04-04 at 18 40 43" src="https://user-images.githubusercontent.com/1613383/161636874-2f6a7ccc-46a7-47eb-9ff1-6b3d2608986a.png">

**Checklist:**

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] I have read the **CONTRIBUTING** document.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.

victorges committed:
api: Split authenticator middleware from the authorizer

victorges committed:
api: Make createdByTokenId an admin-only field

That field basically exposes API keys on all stream
API endpoints. Which means for example that any API
key with access to the /streams API can get access
to any other API key that has ever created streams
(since our API keys are actually the IDs themselves).

This avoids that leak for non-admin users. We should
eventually make the API keys themselves not actually
use the ID as the secret key.

victorges committed:
api: Create separate 401 error different from 403

401 actually means the request is not authenticated (yeah the HTTP name is
    confusing)
403 means that the caller is authenticated but is not allowed to perform
    the action (aka forbidden or unauthorized)

victorges committed:
api: Move CORS and Prometheus to app router

We'll need to add middlewares before CORS now, then
I also didn't want to leave Prometheus all alone in
the root (that test middleware doesnt count).

victorges committed:
api: Validate noApiToken option

victorges committed:
api: Allow API keys to add allowed origins

victorges committed:
api: Add Vary: Authorization to all responses

victorges committed:
api: Fix tests and auth middleware imports

victorges committed:
api: Remove CORS logic from asset controller

hjpotter92 commented on packages/api/src/app-router.ts:
```diff
@@ -35,6 +41,26 @@ const GEOLOCATION_ENDPOINTS = [
+  "https://livepeer.com",
+  "https://livepeer.monster",
+  "https://explorer.livepeer.org",
+  "http://localhost:3000",
```
local port 3004 as well (default for livepeer box)

victorges replied:
Actually the localhost here is unnecessary afaik since browsers already allow cors from localhost by default. I'll remove the 3000 as well.

victorges replied:
Actually, maybe it only worked on localhost becaus we had this here? But probably browsers/the cors library just ignored the port cause I doubt all hackers that ever depended on it ran their applications on port 3000.

Nevertheless, I'll try to improve this list as well. I don't think we should be running production with CORS enabled for localhost. Probably enable it only development and somehow for the box?

victorges replied:
Created the `LP_API_CORS_JWT_ALLOWLIST` which should be set in custom environments where the dashboard is going to be somewhere different from `locahost:3000`. Already set it in the staging environment like this: https://github.com/livepeer/livepeer-infra/blob/f43702e3fdf34adff59090629a76a3411b478e62/values/staging/livepeer.values.yaml#L12

WDYT? Think we can set an env like that in the box with the custom `3004` port?

hjpotter92 replied:
> Created the `LP_API_CORS_JWT_ALLOWLIST` which should be set in custom environments where the dashboard is going to be somewhere different from `locahost:3000`. Already set it in the staging environment like this: [livepeer/livepeer-infra@`f43702e`/values/staging/livepeer.values.yaml#L12](https://github.com/livepeer/livepeer-infra/blob/f43702e3fdf34adff59090629a76a3411b478e62/values/staging/livepeer.values.yaml#L12)

I was concocting the same theory while flying back today ^_^ :+1: a set of csv domains to be populated through environment variables. It accepting regex patterns is super cool :star2: :star_struck: 

> 
> WDYT? Think we can set an env like that in the box with the custom `3004` port?

it should theoretically be possible, but since MistController is the one overtaking spawns, might need to confirm! Maybe @iameli might know right off about this behaviour

victorges committed:
api/cli: Make cors allow-list configured as an arg/env

victorges committed:
api/cors: Make the static allowlist only for JWT reqs

victorges committed:
api/env: Remove localhost from default allowlist

We should still configure prod/staging separately, with
prod only really allowing access from livepeer.com (and maybe explorer?)

victorges committed:
api: Remove explorer from CORS allowlist

Not needed anymore!

Also fix some . that should be \. in the regexes

victorges committed:
api: Create logic for restricting CORS access

Pending actually allowing some APIs to be used

victorges committed:
api: Default CORS to https://${frontendDomain} instead

victorges committed:
api: Allow some stream APIs to use use CORS

victorges committed:
api: Allow some multistream APIs to use CORS

victorges committed:
api: Allow some asset APIs to use CORS

victorges committed:
api: Allow any origin to access /api/asset/upload

victorges committed:
www: Create allow CORS checkbox in API key creation

victorges committed:
www: Add input for allowed origins for API key CORS

victorges committed:
www: Add a + button next to add origin field

victorges committed:
www: Clear API key dialog anytime it is closed

victorges committed:
www: Update copy for less indirection

victorges committed:
www: Allow creating CORS API keys with full API access

victorges committed:
prettier

victorges committed:
www: Fix default suggested CORS origin

victorges committed:
www: Stop disallowing whitespace in token

victorges committed:
www: Remove whitespace from addAllowedOrigin field

victorges committed:
api: Allow any origin in the OPTIONS call

victorges committed:
api: Create getCorsOpts internal helper

Just some cleaner code

victorges committed:
api: Disallow API keys on certain user APIs

Also disallow an user to suspend themelves lol

victorges committed:
api: Fix CORS for API tokens that allow *

victorges committed:
api: Simplify OPTIONS CORS handling

victorges committed:
api: Fix CORS allow-list arg parsing

victorges committed:
api: Make CORS allowed origins list a JSON array

hjpotter92 commented on packages/api/.env:
```diff
@@ -2,7 +2,7 @@ LP_API_BROADCASTERS=[{"address":"http://127.0.0.1:3085","cliAddress":"http://127
 LP_API_JWT_SECRET=secretsecret
 LP_API_JWT_AUDIENCE=livepeer
-LP_API_CORS_JWT_ALLOWLIST=http://localhost:3000
+LP_API_CORS_JWT_ALLOWLIST=["http://localhost:3000"]
```
nit: I don't know if the name `CORS_JWT_ALLOWLIST` is technically correct. CORS is a browser specific limitation/feature, having nothing to do with jwt

victorges replied:
So the JWT part is not from the CORS definition, but from our own implementation of our CORS rules. This allow-list only applies for requests authenticated with a JWT.

If you prefer, I can rename this to `LP_API_JWT_CORS_WHITELIST` instead. Do you think it makes more sense?

hjpotter92 replied:
oh nice! that clears up my doubts. I was just confused with the jwt being present in there.

victorges commented on packages/api/src/controllers/task.ts:
```diff
@@ -171,7 +171,7 @@ app.get("/", authMiddleware({}), async (req, res) => {
 });
 
-app.get("/:id", authMiddleware({}), async (req, res) => {
+app.get("/:id", authorizer({}), async (req, res) => {
```
Gotta allow CORS in this API as well

victorges commented on packages/api/src/controllers/task.ts:
```diff
@@ -205,39 +205,35 @@ app.post(
+});
 
-app.delete("/:id", authMiddleware({}), async (req, res) => {
+app.delete("/:id", authorizer({}), async (req, res) => {
```
This should probably be admin as well

victorges reviewed (commented):
some reminders

victorges commented on packages/api/src/app-router.ts:
```diff
@@ -143,6 +157,25 @@ export default async function makeApp(params: CliArgs) {
+        },
+      })
+    )
+  );
```
There is currently 1 problem with this middleware order:

If an invalid token or JWT is used for the request, the `authenticator` middleware will explode and then the `cors` middleware will never run. This means that if you use an invalid token from the browser you might never even see the response, because the response won't have CORS headers and then the browser will explode the request in a different way for you. 

To fix this, we have to make the `authenticator` middleware never throw an exception. It can be made either by actually never exploding and just letting the `authorizer` middleware explode later with "request not authenticated", or we can do some kind of deferred explode, saving the error in the request to throw after the `cors` middleware. Orr we can also make the CORS an error handler somehow?

I'll explore these possibilites tomorrow.

victorges replied:
Fixed it here! 
https://github.com/livepeer/livepeer-com/pull/985/commits/89d1fac7299d04b9d9e79600976ed7ba798bce6e

victorges reviewed (commented):
Just one other little thing for this to be ready to merge

victorges committed:
api/task: Tune authorizer params

victorges committed:
api: Always run cors middleware after auth

victorges committed:
api: Fix auth middleware

express-async-errors is a lie, we can't really
ignore next for a middleware apparently.

victorges committed:
www: Create column for CORS in API key table

victorges commented on packages/api/src/app-router.ts:
```diff
@@ -143,6 +155,24 @@ export default async function makeApp(params: CliArgs) {
+  app.use(
+    authenticateWithCors({
+      cors: {
+        anyOriginPathPrefixes: [pathJoin("/", httpPrefix, "/asset/upload/")],
```
Hey @anarkrypto! This line here is the one that will make these code changes have exactly the effect that you suggested!

The whole point of the separate `/request-upload` from the `/upload/$token` APIs was that users of the API would not need to proxy the file contents themselves. The `/upload` API doesn't even require authentication, so this behavior of restricting CORS was certainly a bug. 

I've also been testing this PR for the past week and I'm pretty confident that it is good to go. So I'll be merging it tomorrow (just after adding some docs) and will be hopefully deploying it on tuesday! Rest assured that you won't have to deal with file upload traffic from your backend integrating with our VOD API :)

And off course, thanks a lot for your input and detailed issue! If I didn't already have this PR close to merging I'd certainly work on a hotfix just for your use case, knowing you were blocked/affected by this behavior. Cheers!

hjpotter92 reviewed (approved):
lgtm!

iameli reviewed (approved):
Went through all the `{allowCorsApiKey: true}` instances to me and they seem reasonable. And I love the tooltip text. LGTM!

victorges committed:
www: Add origin validation

victorges committed:
www: Keep only * when * is in allowed origins

victorges committed:
www/docs: Add first part of API key CORS docs

victorges committed:
www: Add dummy API access docs and links from creation dialog

victorges committed:
api: Change transcode API to /api/asset/:id/transcode

Still keep backward compatibility since the NFT SDK is already
calling the old one.

victorges committed:
www/docs: Fix API key reference

Made it a little shorter just with usage information
and a link to the getting an API key guide.

victorges committed:
www/docs: Remove domain warning from Stream Health

It is not true anymore!

Also remove the repeated beta badge that messes with the
header. We already have the note right below and it's not
**that** beta anymore lol

victorges committed:
www/docs: Finish API key docs!

victorges committed:
www/docs: Final nits!
